### DISCUSSION TITLE
Vertical integration of shell and terminal emulator

### DISCUSSION THREAD

**Discussion by jan-xyz** [kovidgoyal/kitty#4451] [Ideas]
I spent the last couple of days watching and reading up on terminal emulator, mostly because of [warp.dev](https://www.warp.dev) and it got me thinking: if the vt100 and other archaic protocols are the limiting factor for better developer experience in terminals wouldn't it be possible to completely get rid of the protocol and vertically integrate both shell and terminal into one application? Wouldn't that get rid of the problem that the protocol is in the way? 

Effectively, it would mean that something like kitty provides its own shell, right? Or something like zsh comes with a terminal front end. I know both domains require an insane amount of cognitive load on its own but if two projects would say that they are willing to collaborate more deeply, wouldn't this in turn create a basis to move forward?

**Comment by page-down** [comment]
The shell is only one type of the command line programs. A large number of other programs exist, so it is unrealistic to expect a complete replacement of the current protocol.

If the interactive shell were more tightly integrated with the terminal, it would certainly improve the user experience, but only that part. That is: command execution and result feedback (and others).

Currently it is only possible to move forward with the development of command line interactive terminals while maintaining backward compatibility as much as possible.

This is my opinion.

**Reply by jan-xyz** [reply]
> The shell is only one type of the command line programs.

But don't those other command line programs run _via_ the shell as well? You wouldn't just execute grep or awk on it's own right?

> command execution and result feedback (and others).

That's exactly what I mean "just that part" would already be an improvement over a 30-40 year old protocol which was build for a physical terminal, wouldn't it?

> Currently it is only possible to move forward with the development of command line interactive terminals while maintaining backward compatibility as much as possible.

Maybe focusing on a niche that can "ignore" the other parts would already be a start. Most software development happens on the local machine, no SSH involved, and getting rid of shell scripts or at least for the time being, would probably work for that group as well. I get that in operations you need the ability to SSH into the server sometimes....

**Reply by jan-xyz** [reply]
I think your other reply opened my eyes. I somehow missed the part that the shell is just the entry point and all other CLI programs replace the shell when you execute it until it returns back to shell.

**Reply by page-down** [reply]
@jan-xyz 
For the part where the terminal is tightly integrated with the shell, you might want to take a look at this video. However, I personally don't agree with some of the points made in this experimental project.

https://media.handmade-seattle.com/terminal-click/

**Reply by jan-xyz** [reply]
Oh wow, it's cool to see the parallels to warp and I am curious to see how that plays out. I guess it's exactly in the direction I was thinking of as well. His statement that "the shell has too much power" is probably what lies at the core. It's fine to let it handle process communication, but the whole part that comes before it, e.g. traversing directories what he showed with terminal.click or the very powerful command editor from warp are some really interesting ways to take away that power from the shell and improve the developer experience. Thank you for sharing this talk!

**Comment by SolitudeSF** [comment]
you only run shell in your emulator?

**Reply by jan-xyz** [reply]
yes, do you mean as opposed to via SSH or as script files?

**Reply by page-down** [reply]
Almost every command you type in the shell is a "command line program" and requires the same "protocol" to be used.

For example, vim, grep, ssh, etc.

**Reply by jan-xyz** [reply]
Thanks! That helps 😊

**Comment by kovidgoyal** [comment]
Most of the things in warp are already do-able in a traditional shell, it' just that shell developers dont bother. From their list of headline features:

1) Block I/O (not exactly sure what this is apart from highlighting the output of the previous command if it fails -- this is doable by traditional shells, with some co-operation with terminals using the existing prompt marking protocol, kitty 0.24 itself uses this for its shell integration features

2) Text editing: This again is entirely doable by a traditional shell, it doesnt need anything from the terminal emulator. Whether it's worth doing is a whole other question.

3) Completions: Already done in shells, see fish for canonical example or the various zsh completion plugins

4) Visual history: This can be done in a traditional shell with one simple addition from the terminal emulator, which kitty already pioneered: https://sw.kovidgoyal.net/kitty/unscroll/

Your main sticking point is motivating shell developers to implement these features, not really vertical integration. The terminal+shell ecosystem has been fairly stagnant, thankfully that is changing now-a-days.

**Reply by jan-xyz** [reply]
I see your point that most shells can do those things already, I understood Warp in a way that they reclaim these powers from the shell and bake it into the terminal. So far it seems they just do a lot of eye candy and small things, but it seems like there are bigger plans for the future at least that's what they claim in some of their blog posts and videos.

I somehow I have the gut feeling that they pursue a similar goal to kitty: try to improve the 30 year old status quo.
